veps/vep-4.md

```diff
@@ -0,0 +1,342 @@
+# # VEP-4 Vitess Flag Restructure, `pflag`, and `viper`
+
+```
+VEP: 4
+Title: Vitess Flag Restructure, `pflag`, and `viper`
+Author: Andrew Mason
+Reviewers: Deepthi Sigireddi, Derek Perkins, Richard Bailey
+Status: In Review
+Created: 2021-04-07 (second major revision 2021-11-03)
+```
+
+## Abstract
+
+This VEP suggests changes and outlines a migration path, in phases, to restructure flag use across the Vitess codebase, migrate to [`pflag`][pflag], and eventually to [`viper`][viper].
+
+## Introduction
+
+Vitess's flag usage has evolved over the life of the project into a sort of organic sprawl.
+
+Flags are defined on the global `flag.FlagSet` in the packages that create them — a binary that imports that package (or imports a package which imports a package which imports ... and so on) implicitly gets that flag defined and that flag will appear in the binary's help text [1], whether the flag makes sense or not.
+This impacts the discoverability and ease-of-use of the various binaries.
+
+In addition, this imposes maintenance burdens.
+When a new flag is introduced, care must be taken to avoid name collisions (even if you are adding a flag which is only used by `vttablet`, it may end up importing a flag that's really only used by vtgate but happens to share the same name), and it is difficult to extract small pockets of functionality, because you end up importing all the flags anyway.
+
+These two issues are somewhat entangled with each other, and are some of the primary motivations, in addition to overall binary size, raised in a more general binary structure RFC in [vitessio/vitess#7471].
+
+## tl;dr
+
+1. Deprecate flag usages that are incompatible with `pflag`.
+1. Switch to `pflag`.
+1. Code refactor to encapsulate flag-backed variables.
+1. Adopt `viper`.
+
+## Open Questions
+
+1. The changes in phases 1 and 2 _can_ break existing scripts.
+Details of exactly how and why are described in each phase.
+
+    Users can actually already update the way they pass flags and subcommands in a way that won't break, without any code changes needed in Vitess.
+
+    The question: **Do we need to do a formal deprecation cycle?**
+    Or can we announce in the monthly meeting and in the release notes and proceed with the migration, saving ourselves 3 months of waiting?
+
+## Migration
+
+### Phase 1: Preparation
+
+Unfortunately, we cannot simply drop-in `pflag` as a replacement for `flag`, despite its advertisement to support exactly that.
+
+Since the very beginning, Vitess documentation has encouraged the use of the single-dash form of flags (e.g. `-my_flag`).
```

Comment:
This is the reason we need a deprecation notice. Quite likely everyone is using the single-dash form of flags and changing all of your scripts is non-trivial.
Luckily vitess-operator uses the double-dash form so at least we won't need to update that.

Reply:
My understanding of a deprecation cycle (for example, changing a flag name) is:

1. version 1: old version exists
2. version 2: old and new versions exist, old "marked" (for some definition of marked) as deprecated. this allows people to switch from the old to the new before going to a version where the old won't work
3. version 3: old is removed, only new version exists

For flags, we're already in step 2(...ish) because people can migrate their scripts in their current vitess version. The only thing we're missing is the "marking". If we think it's important that happens in the code (vs in release notes / announcements / monthly call), then I totally get it and we'll do the deprecation notice.